Serialization of a generic array of enum

**Describe the bug**
Fails to serialize a generic array of enum values when it's defined as a parameterized field of a class.

**To Reproduce**

```java
package test.kryo;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.io.Output;
import com.esotericsoftware.minlog.Log;

import java.io.ByteArrayOutputStream;

public class KryoTest {

    public static void main(String[] args) {
        Kryo kryo = new Kryo();
        kryo.setRegistrationRequired(false);

        Log.set(Log.LEVEL_TRACE);

        ArrayEnum<TestEnum> array = new ArrayEnum<>(new TestEnum[]{TestEnum.A});
        Container container = new Container(array);
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        try (Output output = new Output(outputStream)) {
            kryo.writeClassAndObject(output, container);
        }
    }

    enum TestEnum {
        A, B, C;
    }

    public static class ArrayEnum<V extends Enum<V>> {

        private V[] values;

        public ArrayEnum(V[] array) {
            this.values = array;
        }
    }

    public static class Container {

        private ArrayEnum<TestEnum> field;

        public Container(ArrayEnum<TestEnum> array) {
            this.field = array;
        }
    }
}
```

When running this example, Kryo will throw an exception because it will directly try to use an `EnumSerializer` on the array instead of an `ObjectArraySerializer` before.

```
Exception in thread "main" com.esotericsoftware.kryo.KryoException: java.lang.ClassCastException: [Ltest.kryo.KryoTest$TestEnum; cannot be cast to java.lang.Enum
Serialization trace:
values (test.kryo.KryoTest$ArrayEnum)
field (test.kryo.KryoTest$Container)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:101)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:108)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:642)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:70)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:108)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:711)
	at test.kryo.KryoTest.main(KryoTest.java:21)
Caused by: java.lang.ClassCastException: [Ltest.kryo.KryoTest$TestEnum; cannot be cast to java.lang.Enum
	at com.esotericsoftware.kryo.serializers.DefaultSerializers$EnumSerializer.write(DefaultSerializers.java:349)
	at com.esotericsoftware.kryo.Kryo.writeObjectOrNull(Kryo.java:692)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:79)
	... 6 more
```

**Environment:**
 - OS: Ubuntu
 - JDK Version: 8
 - Kryo Version: 5.2.0 and 5.4.0

I've just noticed that it doesn't work for any generic array with a pre-defined serializer. It will produce the same error for `Number`.
```java
package test.kryo;

import com.esotericsoftware.kryo.Kryo;
import com.esotericsoftware.kryo.io.Output;
import com.esotericsoftware.minlog.Log;

import java.io.ByteArrayOutputStream;

public class KryoTest {

    public static void main(String[] args) {
        Kryo kryo = new Kryo();
        kryo.setRegistrationRequired(false);

        Log.set(Log.LEVEL_TRACE);

        ArrayNumber<Integer> array = new ArrayNumber<>(new Integer[]{1});
        Container container = new Container(array);
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        try (Output output = new Output(outputStream)) {
            kryo.writeClassAndObject(output, container);
        }
    }

    public static class ArrayNumber<V extends Number> {

        private V[] values;

        public ArrayNumber(V[] array) {
            this.values = array;
        }
    }

    public static class Container {

        private ArrayNumber<Integer> field;

        public Container(ArrayNumber<Integer> array) {
            this.field = array;
        }
    }
}
```
When running, it will produce the same error:
```
Exception in thread "main" com.esotericsoftware.kryo.KryoException: java.lang.ClassCastException: [Ljava.lang.Integer; cannot be cast to java.lang.Integer
Serialization trace:
values (test.kryo.KryoTest$ArrayNumber)
field (test.kryo.KryoTest$Container)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:101)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:108)
	at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:642)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:70)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:108)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:711)
	at test.kryo.KryoTest.main(KryoTest.java:21)
Caused by: java.lang.ClassCastException: [Ljava.lang.Integer; cannot be cast to java.lang.Integer
	at com.esotericsoftware.kryo.serializers.DefaultSerializers$IntSerializer.write(DefaultSerializers.java:115)
	at com.esotericsoftware.kryo.Kryo.writeObjectOrNull(Kryo.java:692)
	at com.esotericsoftware.kryo.serializers.ReflectField.write(ReflectField.java:79)
	... 6 more

--- a/src/com/esotericsoftware/kryo/serializers/ReflectField.java
+++ b/src/com/esotericsoftware/kryo/serializers/ReflectField.java
@@ -27,6 +27,7 @@ import com.esotericsoftware.kryo.io.Input;
 import com.esotericsoftware.kryo.io.Output;
 import com.esotericsoftware.kryo.serializers.FieldSerializer.CachedField;
 import com.esotericsoftware.kryo.util.Generics.GenericType;
+import com.esotericsoftware.kryo.util.Util;
 
 import java.lang.reflect.Field;
 
@@ -151,7 +152,9 @@ class ReflectField extends CachedField {
 	Class resolveFieldClass () {
 		if (valueClass == null) {
 			Class fieldClass = genericType.resolve(fieldSerializer.kryo.getGenerics());
-			if (fieldClass != null && fieldSerializer.kryo.isFinal(fieldClass)) return fieldClass;
+			if (fieldClass != null && fieldSerializer.kryo.isFinal(fieldClass)) {
+				return field.getType().isArray() ? Util.getArrayType(fieldClass) : fieldClass;
+			}
 		}
 		return valueClass;
 	}

--- a/src/com/esotericsoftware/kryo/util/Util.java
+++ b/src/com/esotericsoftware/kryo/util/Util.java
@@ -27,6 +27,7 @@ import com.esotericsoftware.kryo.serializers.ClosureSerializer;
 import com.esotericsoftware.kryo.serializers.FieldSerializer;
 import com.esotericsoftware.kryo.util.Generics.GenericType;
 
+import java.lang.reflect.Array;
 import java.lang.reflect.Type;
 import java.util.HashMap;
 import java.util.Map;
@@ -108,6 +109,21 @@ public class Util {
 		if (type == Short.class) return short.class;
 		if (type == Void.class) return void.class;
 		return type;
+	}
+
+	/** Returns the array type for a given class */
+	public static Class getArrayType(Class type) {
+		if (type == String.class) return String[].class;
+		if (type == Integer.class) return Integer[].class;
+		if (type == Float.class) return Float[].class;
+		if (type == Boolean.class) return Boolean[].class;
+		if (type == Byte.class) return Byte[].class;
+		if (type == Long.class) return Long[].class;
+		if (type == Character.class) return Character[].class;
+		if (type == Double.class) return Double[].class;
+		if (type == Short.class) return Short[].class;
+		// See Class#arrayType() available from JDK 12
+		return Array.newInstance(type, 0).getClass();
 	}
 
 	public static boolean isWrapperClass (Class type) {

--- a/test/com/esotericsoftware/kryo/serializers/GenericsTest.java
+++ b/test/com/esotericsoftware/kryo/serializers/GenericsTest.java
@@ -145,6 +145,38 @@ class GenericsTest extends KryoTestCase {
 		roundTrip(70, o);
 	}
 
+	// Test for https://github.com/EsotericSoftware/kryo/issues/940
+	@Test
+	void testFieldWithStringArrayType () {
+		StringArray array = new StringArray(new String[] {"1"});
+		
+		kryo.setRegistrationRequired(false);
+
+		roundTrip(67, array);
+	}
+
+	// Test for https://github.com/EsotericSoftware/kryo/issues/940
+	@Test
+	void testFieldWithNumberArrayType () {
+		NumberArray<Integer> array = new NumberArray<>(new Integer[] {1});
+		NumberArrayHolder container = new NumberArrayHolder(array);
+		
+		kryo.setRegistrationRequired(false);
+
+		roundTrip(137, container);
+	}
+
+	// Test for https://github.com/EsotericSoftware/kryo/issues/940
+	@Test
+	void testFieldWithObjectArrayType () {
+		ObjectArray<TestObject> array = new ObjectArray<>(new TestObject[] {new TestObject(1)});
+		ObjectArrayHolder container = new ObjectArrayHolder(array);
+		
+		kryo.setRegistrationRequired(false);
+
+		roundTrip(265, container);
+	}
+
 	// Test for https://github.com/EsotericSoftware/kryo/issues/655
 	@Test
 	void testClassWithMultipleGenericTypes() {
@@ -565,4 +597,116 @@ class GenericsTest extends KryoTestCase {
 		}
 	}
 
+	public static class StringArray {
+
+		private String[] values;
+
+		public StringArray() {
+		}
+
+		public StringArray(String[] array) {
+			this.values = array;
+		}
+
+		public boolean equals(Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			StringArray that = (StringArray) o;
+			return Arrays.equals(values, that.values);
+		}
+	}
+
+	public static class NumberArray<V extends Number> {
+
+		private V[] values;
+
+		public NumberArray() {
+		}
+
+		public NumberArray(V[] array) {
+			this.values = array;
+		}
+
+		public boolean equals(Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			NumberArray<?> that = (NumberArray<?>) o;
+			return Arrays.equals(values, that.values);
+		}
+	}
+
+	public static class NumberArrayHolder {
+
+		private NumberArray<Integer> field;
+
+		public NumberArrayHolder() {
+		}
+
+		public NumberArrayHolder(NumberArray<Integer> array) {
+			this.field = array;
+		}
+
+		public boolean equals(Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			NumberArrayHolder that = (NumberArrayHolder) o;
+			return Objects.equals(field, that.field);
+		}
+	}
+
+	public static class ObjectArray<V> {
+
+		private V[] values;
+
+		public ObjectArray() {
+		}
+
+		public ObjectArray(V[] array) {
+			this.values = array;
+		}
+
+		public boolean equals (Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			ObjectArray<?> that = (ObjectArray<?>)o;
+			return Arrays.equals(values, that.values);
+		}
+	}
+
+	public static class ObjectArrayHolder {
+
+		private ObjectArray<TestObject> field;
+
+		public ObjectArrayHolder() {
+		}
+
+		public ObjectArrayHolder(ObjectArray<TestObject> array) {
+			this.field = array;
+		}
+
+		public boolean equals (Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			ObjectArrayHolder that = (ObjectArrayHolder)o;
+			return Objects.equals(field, that.field);
+		}
+	}
+
+	public static class TestObject {
+		private int i;
+
+		public TestObject() {
+		}
+
+		public TestObject(int i) {
+			this.i = i;
+		}
+
+		public boolean equals(Object o) {
+			if (this == o) return true;
+			if (o == null || getClass() != o.getClass()) return false;
+			TestObject that = (TestObject) o;
+			return i == that.i;
+		}
+	}
 }

--- a/test/com/esotericsoftware/kryo/util/UtilTest.java
+++ b/test/com/esotericsoftware/kryo/util/UtilTest.java
@@ -44,4 +44,12 @@ class UtilTest {
         assertFalse(Util.isAssignableTo(String.class, long.class));
     }
 
+    @Test
+    void testGetArrayType() {
+        assertEquals(int[].class, Util.getArrayType(int.class));
+        assertEquals(Integer[].class, Util.getArrayType(Integer.class));
+        assertEquals(String[].class, Util.getArrayType(String.class));
+        assertEquals(String[][].class, Util.getArrayType(String[].class));
+        assertEquals(Object[].class, Util.getArrayType(Object.class));
+    }
 }

